captum/attr/_core/feature_ablation.py

```diff
@@ -362,21 +360,43 @@ def attribute(
                     if show_progress:
                         attr_progress.update()
 
-                    if not agg_output_mode:
-                        # current_batch_size is not n_examples
-                        # it may get expanded by n_feature_perturbed
+                    # if perturbations_per_eval > 1, the output shape must grow with
+                    # input and not be aggregated
+                    if perturbations_per_eval > 1 and not self._is_output_shape_valid:
                         current_batch_size = current_inputs[0].shape[0]
+
+                        # number of perturbation, which is not the same as
+                        # perturbations_per_eval when not enough features to perturb
+                        n_perturb = current_batch_size / num_examples
+
+                        current_output_shape = modified_eval.shape
+
+                        # use initial_eval as the forward of perturbations_per_eval = 1
+                        initial_output_shape = initial_eval.shape
+
                         assert (
-                            modified_eval.numel() == current_batch_size
-                        ), """expected output of forward_func to grow with
-                        batch_size. If this is not the case for your model
-                        please set perturbations_per_eval = 1"""
+                            # check if the output is not a scalar
+                            current_output_shape
+                            and initial_output_shape
+                            # check if the output grow in same ratio, i.e., not agg
+                            and current_output_shape[0]
+                            == n_perturb * initial_output_shape[0]
+                        ), (
+                            "When perturbations_per_eval > 1, forward_func's output "
+                            "should be a tensor whose 1st dim grow with the input "
+                            f"batch size: when input batch size is {num_examples}, "
```

Comment:
I think we can be more strict here by requiring the 1st dim of output to be the same as input batch size, instead of just require it "grow with the input batch size".

For example, if the input batch is `2`, the output's 1st dim can be `2`, `4`, or `7`; as long as
when we expand the input batch size to `4`, the output's 1st dim becomes `4`, `8`, or `14` respectively. 

But the last 2 cases are pretty weird and unlikely to happen. It may be easier to always require output's 1st dim to be batch size and it must be the same as input batch size in this "non-aggregation mode"